### PR TITLE
Attempt to gather kubelet logs from workers even if they never made it to a Node

### PR DESCRIPTION
**Pull request by dgoodwin**
- Require a ssh bastion for azure 4.9 to 4.10 upgrade jobs.
- Temporary attempt to gather worker kubelet logs even if no node exists.

### CODE REVIEW

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh [hunk 1]
```diff
@@ -271,6 +271,23 @@ echo "$(date) - waiting for clusteroperators to finish progressing..."
...
+	# We should have an ssh bastion pod per above. Now we use it to try to gather kubelet logs from each worker.
+	WORKER_IPS="$(oc --insecure-skip-tls-verify get machines -n openshift-machine-api -l machine.openshift.io/cluster-api-machine-role=worker -o 'jsonpath={.items[*].status.addresses[?(@.type=="InternalIP")].address}')"
+	echo "Found worker IPs: $WORKER_IPS"
+	ingress_host="$(oc get service --all-namespaces -l run=ssh-bastion -o go-template='{{ with (index (index .items 0).status.loadBalancer.ingress 0) }}{{ or .hostname .ip }}{{end}}')"
+	echo "Ingress host: $ingress_host"
+
+	for ip in $WORKER_IPS
+	do
+		echo "Gathering kubelet logs from ${ip}"
+	    	ssh -i "${KUBE_SSH_KEY_PATH}" -t -o StrictHostKeyChecking=no -o ProxyCommand="ssh -i ${KUBE_SSH_KEY_PATH} -A -o StrictHostKeyChecking=no -o ServerAliveInterval=30 -W %h:%p core@${ingress_host}" core@$ip "sudo journalctl -u kubelet --no-pager" > "${ARTIFACT_DIR}/worker-kubelet-logs/${ip}-kubelet.log"
```
Comment (JoelSpeed) [inline ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:287]: IIRC, when we gather logs on nodes via must gather, we don't filter by unit normally. If we filter to the kubelet unit, any dependency failures won't be picked up. Is it better to just take the whole journal and filter manually later?

Comment (dgoodwin) [reply ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:287]: Yeah that seems fine to me, I'll switch it over.

**Comment by dgoodwin** [comment]
I think this is good for review now. You can see a sample here: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/23640/rehearse-23640-periodic-ci-openshift-release-master-ci-4.10-upgrade-from-stable-4.9-e2e-azure-upgrade/1460622658256244736/artifacts/e2e-azure-upgrade/openshift-e2e-test/artifacts/machine-journal-logs/

I've now modified to only run when we detect missing nodes. Once merged we should have data within a day.

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh [hunk 2]
```diff
@@ -275,20 +275,40 @@ echo "$(date) - all clusteroperators are done progressing."
...
+        break
+    fi
+    echo "$(date) - $MACHINECOUNT Machines - $NODECOUNT Nodes"
+    sleep 30
+    i=$((i+1))
+    if [ $i -gt 20 ]; then
+        echo "Timed out waiting for node count ($NODECOUNT) to equal or exceed machine count ($MACHINECOUNT)."
+        # If we enabled the ssh bastion pod, attempt to gather journal logs from each machine, regardless
+        # if it made it to a node or not.
+        # TODO: Temporary, move into the failure case block below before merging:
```
Comment (JoelSpeed) [inline ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:292]: Is this TODO done now?

**Comment by deads2k** [comment]
/approve

but leaving detailed review to those more bash knowledgeable

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh [hunk 3]
```diff
@@ -275,20 +275,40 @@ echo "$(date) - all clusteroperators are done progressing."
 i=0
```
Comment (deads2k) [inline ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:275]: please move the clusteroperator wait *after* the nodes ready check please

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh (hunk 2 above)
Comment (dgoodwin) [reply ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:292]: My bad yes it's done, will remove.

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh (hunk 3 above)
Comment (dgoodwin) [reply ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:269]: done

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh [hunk 4]
```diff
@@ -265,30 +265,43 @@ oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.8-kube-1.22-
...
+            # This returns each IP all on one line, separated by spaces:
+            machine_ips="$(oc --insecure-skip-tls-verify get machines -n openshift-machine-api -o 'jsonpath={.items[*].status.addresses[?(@.type=="InternalIP")].address}')"
+            echo "Found machine IPs: $machine_ips"
+            ingress_host="$(oc get service --all-namespaces -l run=ssh-bastion -o go-template='{{ with (index (index .items 0).status.loadBalancer.ingress 0) }}{{ or .hostname .ip }}{{end}}')"
+            echo "Ingress host: $ingress_host"
+
+            for ip in $machine_ips
+            do
+                echo "Gathering journalctl logs from ${ip}"
+                ssh -i "${KUBE_SSH_KEY_PATH}" -o StrictHostKeyChecking=no -o ProxyCommand="ssh -i ${KUBE_SSH_KEY_PATH} -A -o StrictHostKeyChecking=no -o ServerAliveInterval=30 -W %h:%p core@${ingress_host}" core@$ip "sudo journalctl --no-pager" > "${ARTIFACT_DIR}/machine-journal-logs/${ip}-journal.log"
```
Comment (dgoodwin) [inline ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:299]: This magic is based on https://github.com/eparis/ssh-bastion

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh [hunk 5]
```diff
@@ -297,6 +310,12 @@ echo "$(date) - waiting for nodes to be ready..."
 oc wait nodes --all --for=condition=Ready=true --timeout=10m
 echo "$(date) - all nodes are ready"
 
+# wait for all clusteroperators to reach progressing=false to ensure that we achieved the configuration specified at installation
+# time before we run our e2e tests.
+echo "$(date) - waiting for clusteroperators to finish progressing..."
+oc wait clusteroperators --all --for=condition=Progressing=false --timeout=10m
+echo "$(date) - all clusteroperators are done progressing."
```
Comment (dgoodwin) [inline ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:317]: Moved to after Nodes checks per @deads2k  request

Comment (deads2k) [reply ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:317]: > Moved to after Nodes checks per @deads2k request

Thanks. some operators have work to do *after* nodes are ready, so we want to wait for them to finish.  This looks correct to me.

**Pull request merged**